docs/src/proposals/fee_transaction_priority.md

```diff
@@ -0,0 +1,210 @@
+    match blocked_transaction_entry {
+        Occupied(existing_blocked_transaction) => {
+            // If there is already a set of transactions blocked on this account, add
+            // this transaction to the priority queue.
+            existing_blocked_transaction.insert_transaction(next_highest_transaction);
+        }
+
+        Vacant(vacant_entry) => {
+            // Create a new queue blocked on this transaction
+            let new_blocked_transaction_queue =
+                Rc::new(BlockedTransactionsQueue {
+                    highest_priority_blocked_transaction: next_highest_transaction,
+                    other_blocked_transactions: BinaryHeap::new(),
+                });
+            // Insert into the hashmap for this `account_key`
+            vacant_entry.insert(new_blocked_transaction_queue.clone());
+            // Insert into the `blocked_transactions` hashmap to indicate this set of transactions
+            // is blocked by `next_highest_transaction`
+            self.blocked_transactions.insert(
+                next_highest_transaction.signature(),
+                new_blocked_transaction_queue
+            );
+        }
+    }
+}
+```
+
+5. Run until all `N` BankingStage threads have been sent `processing_batch` transactions (i.e. hit step 3 above).
+
+#### Banking Threads
+1. Banking threads maintain a queue of transactions sent to them by the scheduler, sorted by priority.
+2. Because the scheduler has guaranteed that there are no locking conflicts, the banking thread can process
+some `M` of these transactions at a time and pack them into entries
+
+#### Handling Completion Signals from BankingStage Threads
+
+Outside of the main loop above, we rely on BankingThreads threads signaling us they've finished their
+task to schedule the next transactions.
+
+1. Once a BankingStage thread finishes processing a batch of transactions `completed_transactions_batch` ,
+it sends the `completed_transactions_batch` back to the scheduler via the same channel to signal of completion.
+
+2. Upon receiving this signal, the BankingStage thread processes the locked accounts
+`transaction_accounts` for each `completed_transaction` in `completed_transactions_batch`:
+```
+let mut unlocked_accounts = vec![];
+// First remove all the locks from the tracking list
+for locked_account in transaction_accounts {
+    if self.locked_accounts.remove_reference(locked_account) {
+        unlocked_accounts.push(locked_account.key());
+    }
+}
+
+// Check if freeing up these accounts has now allowed any new
+// blocked transactions to run
+for account_key in unlocked_accounts {
+    if let Some(blocked_transaction_queue) = self.blocked_transaction_queues_by_accounts.get(account_key) {
+        // Check if the transaction blocking this queue can be run now, thereby unblocking this queue
+        if blocked_transaction_queue.highest_priority_blocked_transaction.can_get_locks() {
```

Comment:
Instead of unblocking only the top of queue, you can unblock the first `k` elements to potentially avoid starvation. After processing you can also allow any of the first `q` elements of the global blocked queue to skip the line if they are now freed up.

Reply:
We discussed separately, but it probably also makes sense to keep track of the number of times that the leader was cut, as to prevent the highest prio tx from also getting starved